### PR TITLE
Added Visceroid fusing

### PR DESCRIPTION
**Pull request by Mailaender**
Closes https://github.com/OpenRA/OpenRA/issues/10791.

For generating a lot of small test visceroids I suggest the following change to defaults.yaml:
```diff
@@ -584,7 +584,7 @@
 		CrushSound: squish6.aud
 	SpawnActorOnDeath:
 		Actor: visc_sml
-		Probability: 10
+		Probability: 100
 		OwnerType: InternalName
 		InternalOwner: Creeps
 		DeathType: TriggerVisceroid
```

### CODE REVIEW

**Comment by GraionDilach** [comment]
Note that this method is the most accurate one.

**Comment by pchote** [comment]
This is a clever and simple solution, so I would be tempted to go with this even if it weren't the most accurate approach.

**Comment by Mailaender** [comment]
Yep: KISS principle here. I assume for visceroids you won't notice as baby visceroids can't fire and the adult ones have a very short range and their shoot/move sequence look very much alike.

**Comment by pchote** [comment]
Other way around: things firing *on* the actor.

**Comment by Mailaender** [comment]
I see. That will be noticeable, but most of the visceroid encounters will be random and auto target triggered anyway. I also wonder how likely it is that random wandering visceroids fuse. For my tests I used unrealistic numbers of infantry Tiberium inflicted infantry deaths and also build a wall around the Tiberium field so they don't wander away from each other.

**Review by pchote — DISMISSED** [review]

File: OpenRA.Mods.Common/Traits/TransformOnCrush.cs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+using System.Collections.Generic;
+using OpenRA.Mods.Common.Activities;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public class TransformOnCrushInfo : ITraitInfo
+	{
+		[ActorReference] public readonly string IntoActor = null;
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/TransformOnCrush.cs]: Hm, do we want this to have a `Fieldloader.Require` (we're not very consistent with that)?

Comment (Mailaender) [reply OpenRA.Mods.Common/Traits/TransformOnCrush.cs]: Good idea. I assume it will crash if not set. Done.

**Comment by abcdefg30** [comment]
Oh, seems like the transport helicopter husk is broken. Mind to fix that as well here?
```
OpenRA.Utility(1,1): Error: Failed with exception: System.AggregateException: One or more errors occurred. ---> OpenRA.YamlException: Actor type tran.husk1: Trait name TransformOnCapture: Required property missing: IntoActor
```

**Comment by Mailaender** [comment]
Sorry, I didn't `make test` as it takes very long on my notebook. #13937

File: OpenRA.Mods.Common/Traits/TransformOnCrush.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+		[ActorReference, FieldLoader.Require] public readonly string IntoActor = null;
+
+		public readonly bool SkipMakeAnims = true;
+
+		public readonly HashSet<string> CrushClasses = new HashSet<string>();
+
+		public virtual object Create(ActorInitializer init) { return new TransformOnCrush(init, this); }
+	}
+
+	public class TransformOnCrush : INotifyCrushed
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/TransformOnCrush.cs]: The trait name suggests that the crushed actor will transform into something else, but it actually force-transforms the *crusher*.

File: OpenRA.Mods.Common/Traits/TransformOnCrush.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+			if (!info.CrushClasses.Overlaps(crushClasses))
+				return;
+
+			var facing = crusher.TraitOrDefault<IFacing>();
+			var transform = new Transform(crusher, info.IntoActor) { Faction = faction };
+			if (facing != null)
+				transform.Facing = facing.Facing;
+
+			transform.SkipMakeAnims = info.SkipMakeAnims;
+			crusher.CancelActivity();
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/TransformOnCrush.cs]: If this returns false then queuing the transform would be bogus.

File: OpenRA.Mods.Common/Traits/TransformOnCrush.cs (hunk 2 above)
Comment (Mailaender) [reply OpenRA.Mods.Common/Traits/TransformOnCrush.cs]: What do you suggest instead? Seems pretty clear to me.

Comment (pchote) [reply OpenRA.Mods.Common/Traits/TransformOnCrush.cs]: `TransformCrusherOnCrush` would be the simplest option.

**Comment by Mailaender** [comment]
Updated.

**Comment by reaperrr** [comment]
Needs rebase.

**Comment by Mailaender** [comment]
Rebased.

**Comment by Arular101** [comment]
Please don't forget to update the copyright notice year.

**Comment by Mailaender** [comment]
Updated the copyright year.

**Review by reaperrr — APPROVED** [review]

**Review by GraionDilach — APPROVED** [review]
:+1:

**Pull request merged**